neurodamus/utils/memory.py

```diff
@@ -568,3 +577,91 @@ def validate_inputs_distribute(self, num_ranks, batch_size):
+            for rank_id in range(num_ranks):
+                for cycle_id in range(cycles):
+                    if not rank_allocation.get((rank_id, cycle_id)):
+                        logging.warning(f"Bucket ({rank_id}, {cycle_id}) in population '{pop}' "
+                                        f"has no GIDs assigned.")
+                        return False
+        return True
+
+    @run_only_rank0
+    def distribute_cells_with_validation(self,
+                                         num_ranks,
+                                         cycles=None,
+                                         metype_file=None
+                                         ) -> Tuple[dict, dict, dict]:
+        """
+        Wrapper function to distribute cells with ever smaller assignment batches until
+        all buckets have at least one GID assigned (or fail if no valid distribution is found).
+
+        Args:
+            num_ranks (int): The number of ranks.
+            cycles (int): The number of cycles to distribute cells over.
+            metype_file (str): The path to a JSON file containing memory usage for each METype.
+
+        Returns:
+            Tuple[dict, dict, dict]: Returns the same as distribute_cells once a valid distribution
+                                     is found.
+        """
+        if not cycles:
+            cycles = 1
+        logging.info("Distributing cells across %d ranks and %d cycles", num_ranks, cycles)
+
+        valid_distribution = False
+
+        def _calculate_total_elements_per_population(self):
+            total_elements_per_population = {}
+            for population, metypes in self.pop_metype_gids.items():
+                total_elements = sum(len(gids) for gids in metypes.values())
+                total_elements_per_population[population] = total_elements
+            return total_elements_per_population
+
+        total_cells_per_population = _calculate_total_elements_per_population(self)
+        average_cells_per_population = {
+            population: total / (num_ranks * cycles)
+            for population, total in total_cells_per_population.items()}
+
+        batch_size = {
+            population: max(1, int(average / MAX_ALLOCATION_STEPS))
+            for population, average in average_cells_per_population.items()}
+
+        while not valid_distribution and any(size > 0 for size in batch_size.values()):
+            bucket_allocation, bucket_memory, metype_memory_usage = self.distribute_cells(
+                num_ranks, cycles, metype_file, batch_size=batch_size
+            )
+            valid_distribution = self.check_all_buckets_have_gids(bucket_allocation,
+                                                                  num_ranks,
+                                                                  cycles)
+            if not valid_distribution:
+                batch_size = {population: max(0, size - 1)
+                              for population, size in batch_size.items()}
```

Comment:
It checks that every rank should contain at least 1 gid from every population, am I reading it correctly? If in the multiple-population case we have population A with a lot of cells, population B with a few cells, to distribute pop B evenly, we also need to reduce the batch size for pop A and redistribute pop A each time?

Reply:
Well spotted Weina! 
To address this, I can modify the logic to reduce the batch size only for the populations that are causing the imbalance (i.e., populations with fewer cells). This way, I can avoid unnecessarily redistributing populations with many cells.

Thanks a lot for the comment!

Reply:
@WeinaJi I've fixed the issue with the latest commit. Please give it a look when you have time.